### PR TITLE
add particle identifier, parent identifier and particle tracking runScript

### PR DESCRIPTION
**Pull request by matthiasto**
the particles dict now has extra information: - ID unique identifier for each particle
                                                                      -parentID (tracks the parents when splitting particles

This enables us to follow particle in time and extract their properties.

run runCom1DFA.py saving particles  (with entrainment on avaAlr)
To visualize the particle tracking, you need to activate it and give in the required info in TRACKPARTICLES section of the ini file

### CODE REVIEW

**Review by awirb — CHANGES_REQUESTED** [review]
could you split the mainTrackParticles so that the trackedParticles flag can be added in the savePartToCsv

**Review by awirb — CHANGES_REQUESTED** [review]
couldn't the runTrackParticles be an option and if flag True be performed in the runCom1DFA.py too? as most of the input would already be available there already - so that it can also be done in the main run if desired?

File: avaframe/com1DFA/DFAtools.py [hunk 1]
```diff
@@ -392,16 +427,133 @@ def mergeParticleDict(particles1, particles2):
         if key == 'Npart':
             particles['Npart'] = particles1['Npart'] + particles2['Npart']
         elif (key in particles2) and (type(particles1[key]).__module__ == np.__name__):
-            if np.size(particles1[key]) == nPart1:
+            if (key == 'ID') or (key == 'parentID'):
```
Comment (fso42) [inline avaframe/com1DFA/DFAtools.py:430]: Please add a few short comment on what's happening here

File: avaframe/com1DFA/DFAtools.py [hunk 2]
```diff
@@ -501,3 +653,4 @@ def scalProd(ux, uy, uz, vx, vy, vz):
     scal = ux*vx + uy*vy + uz*vz
 
     return scal
+    return scal
```
Comment (fso42) [inline avaframe/com1DFA/DFAtools.py:656]: dublicate line?

File: avaframe/com1DFA/com1DFA.py [hunk 3]
```diff
@@ -1700,6 +1763,53 @@ def readPartFromPickle(inDir, flagAvaDir=False):
     return Particles, TimeStepInfo
 
 
+def mainTrackParticles(ParticlesList, TimeStepInfo, center, radius, particleProperties):
```
Comment (fso42) [inline avaframe/com1DFA/com1DFA.py:1766]: - TimeStepInfo is missing from the docstring
- please rename variable to particleList and timeStepInfo

File: avaframe/com1DFA/com1DFA.py [hunk 4]
```diff
@@ -1920,3 +2039,4 @@ def getSimTypeList(simTypeList, inputSimFiles):
             raise FileNotFoundError(message)
 
     return simTypeList
+    return simTypeList
```
Comment (fso42) [inline avaframe/com1DFA/com1DFA.py:2042]: duplicate line?

File: avaframe/com1DFA/runTrackParticles.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
+import pathlib
```
Comment (fso42) [inline avaframe/com1DFA/runTrackParticles.py:1]: This file should live in runScripts, no?

**Review by fso42 — CHANGES_REQUESTED** [review]

File: avaframe/com1DFA/DFAtools.py [hunk 6]
```diff
@@ -5,6 +5,7 @@
 # Load modules
 import logging
 import numpy as np
+import numbers
```
Comment (fso42) [inline avaframe/com1DFA/DFAtools.py:8]: where does numbers come from? Does it need to go to the requirements?

File: avaframe/com1DFA/DFAtools.py [hunk 7]
```diff
@@ -388,20 +423,166 @@ def mergeParticleDict(particles1, particles2):
...
+
+
+def getTrackedParticlesProperties(particlesList, nPartTracked, properties):
+    '''Get the desired properties for the tracked particles
+
+    Parameters
+    ----------
+    particlesList : list
+        list of particles dictionaries (with the 'parentID' array)
+    TimeStepInfo : list
```
Comment (fso42) [inline avaframe/com1DFA/DFAtools.py:547]: This one is not in the parameter list?

File: avaframe/com1DFA/plotTrackParticles.py [hunk 8]
```diff
@@ -0,0 +1,93 @@
+import numpy as np
```
Comment (fso42) [inline avaframe/com1DFA/plotTrackParticles.py:1]: I think it is not a good idea to have the plot show (i.e. popup) by default. I think the default should be to plot to file and only show the plot if it was specifically requested

**Review by fso42 — CHANGES_REQUESTED** [review]

File: avaframe/com1DFA/plotTrackParticles.py [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+import matplotlib.pyplot as plt
+from matplotlib.animation import FuncAnimation, PillowWriter
+
+# Local imports
+import avaframe.com1DFA.DFAtools as DFAtls
+import avaframe.in3Utils.geoTrans as geoTrans
+import avaframe.out3Plot.plotUtils as pU
+
+
+def plotTrackParticle(Particles, trackedPartProp, cfgTrackPart, demOri, dem):
```
Comment (awirb) [inline avaframe/com1DFA/plotTrackParticles.py:11]: move plot function to out3Plots, add title, make m, fd, velocity and x,y plot

File: avaframe/in3Utils/initializeProject.py [hunk 10]
```diff
@@ -29,9 +30,10 @@ def _checkAvaDirVariable(avaDir):
 
     # make sure avaDir is a string
     isString = isinstance(avaDir, str)
-    if not isString:
-        log.warning("AvaDir is not a string, returning")
-        return 'AvaDir is NOT a string'
+    isPurePath = isinstance(avaDir, pathlib.PurePath)
+    if not ((isString) or (isPurePath)):
+        log.warning("AvaDir is not a string or a PurePath, returning")
+        return 'AvaDir is NOT a string or PurePath'
```
Comment (awirb) [inline avaframe/in3Utils/initializeProject.py:36]: maybe better use avaDir not AvaDir

File: avaframe/tests/test_com1DFA.py [hunk 11]
```diff
@@ -1236,3 +1346,38 @@ def test_initializeSimulation():
     assert reportAreaInfo['entrainment'] == 'Yes'
     assert reportAreaInfo['resistance'] == 'No'
     assert np.sum(particles2['secondaryReleaseInfo']['rasterData']) == 4.5
+
+
+def test_runCom1DFA():
```
Comment (awirb) [inline avaframe/tests/test_com1DFA.py:1351]: add tmp_path copy data there to run in a tmp dir

**Review by awirb — CHANGES_REQUESTED** [review]

File: avaframe/com1DFA/DFAfunctionsCython.pyx [hunk 12]
```diff
@@ -911,7 +911,6 @@ def updateFieldsC(cfg, particles, dem, fields):
     hbb = getScalar(Lx0, Ly0, w[0], w[1], w[2], w[3], FDBilinear)
     hBB[j] = hbb
 
-  particles['hBilinearBilinear'] = np.asarray(hBB)
```
[the remark this answers is not in the document]
Comment (matthiasto) [reply avaframe/com1DFA/DFAfunctionsCython.pyx:914]: it was not used anymore, I had it at the beginning when we tried those hBilinBilin or hNearesBilin or hSPH.
We now use by default hBilinBilin and it is called simply h

File: avaframe/com1DFA/DFAtools.py (hunk 6 above)
Comment (matthiasto) [reply avaframe/com1DFA/DFAtools.py:8]: should be already installed. I use this to check if a value is a float, int double... or a string or other none numeric values

File: avaframe/com1DFA/com1DFACfg.ini [hunk 13]
```diff
@@ -182,6 +182,18 @@ entDefResistance = 0
 # will inclune the points located on the polygon line)
 thresholdPointInPoly = 0.001
 
+[TRACKPARTICLES]
+# if particles should be tracked - requires to save particles in OUTPUTS
+trackParticles = False
+# center of the location of the particles to track (x|y coordinates)
+center = 2933|-4010
```
[the remark this answers is not in the document]
Comment (matthiasto) [reply avaframe/com1DFA/com1DFACfg.ini:189]: why that?

**Review by awirb — CHANGES_REQUESTED** [review]

File: avaframe/com1DFA/com1DFACfg.ini [hunk 14]
```diff
@@ -182,6 +182,17 @@ entDefResistance = 0
 # will inclune the points located on the polygon line)
 thresholdPointInPoly = 0.001
 
+[TRACKPARTICLES]
+# if particles should be tracked - requires to save particles in OUTPUTS
```
Comment (fso42) [inline avaframe/com1DFA/com1DFACfg.ini:186]: ist it 'requires' or 'required' ?

File: avaframe/com1DFA/com1DFA.py [hunk 15]
```diff
@@ -327,16 +346,21 @@ def reportAddTimeMassInfo(reportDict, tcpuDFA, infoDict):
...
 
     # add mass info
-    reportDict['Simulation Parameters'].update({'Initial mass [kg]': ('%.2f' % infoDict['initial mass'])})
-    reportDict['Simulation Parameters'].update({'Final mass [kg]': ('%.2f' % infoDict['final mass'])})
-    reportDict['Simulation Parameters'].update({'Entrained mass [kg]': ('%.2f' % infoDict['entrained mass'])})
-    reportDict['Simulation Parameters'].update({'Entrained volume [m3]': ('%.2f' % infoDict['entrained volume'])})
+    reportDict['Simulation Parameters'].update(
+        {'Initial mass [kg]': ('%.2f' % infoDict['initial mass'])})
+    reportDict['Simulation Parameters'].update(
+        {'Final mass [kg]': ('%.2f' % infoDict['final mass'])})
```
Comment (fso42) [inline avaframe/com1DFA/com1DFA.py:352]: There's now a ton of line breaks which are way below our 120 characters. Could it be that your editor is set to 80 character line breaks? (i am not commenting on all these lines...)

File: avaframe/in3Utils/initializeProject.py [hunk 16]
```diff
@@ -67,8 +69,8 @@ def cleanModuleFiles(avaDir, module, alternativeName=''):
         modName = os.path.splitext(modName)[0]
 
     # output dir
-    outDir =  os.path.join(avaDir, 'Outputs')
-    workDir =  os.path.join(avaDir, 'Work')
+    outDir = os.path.join(avaDir, 'Outputs')
+    workDir = os.path.join(avaDir, 'Work')
```
Comment (fso42) [inline avaframe/in3Utils/initializeProject.py:73]: switch this to pathlib?

File: avaframe/out3Plot/outCom1DFA.py [hunk 17]
```diff
@@ -0,0 +1,131 @@
+import numpy as np
+import matplotlib.pyplot as plt
+from matplotlib.animation import FuncAnimation, PillowWriter
```
Comment (fso42) [inline avaframe/out3Plot/outCom1DFA.py:3]: Please include docstrings for the functions

File: avaframe/out3Plot/plotUtils.py [hunk 18]
```diff
@@ -326,7 +334,7 @@ def NonUnifIm(ax, x, y, z, xlab, ylab, **kwargs):
     return ref, im
 
 
-def saveAndOrPlot(pathDict, cfgFlags, outFileName, fig):
+def saveAndOrPlot(pathDict, outFileName, fig):
```
Comment (fso42) [inline avaframe/out3Plot/plotUtils.py:337]: Same here, please update/complete the docstrings

**Review by fso42 — CHANGES_REQUESTED** [review]

**Review by fso42 — APPROVED** [review]

**Pull request merged**